src/axom/core/tests/core_benchmark_array.cpp

```diff
@@ -0,0 +1,528 @@
+  {
+    Container arr;
+    arr.reserve(size);
+    assert_size_and_strict_capacity(arr, 0, size);
+    for(int i = 0; i < size; ++i)
+    {
+      arr.push_back(get_value<T>(i));  // using push_back on container w/ reserved size
+    }
+    assert_size_and_strict_capacity(arr, size, size);
+    benchmark::DoNotOptimize(arr);
+  }
+}
+
+template <typename Container>
+void emplace_back_initialReserve(benchmark::State& state)
+{
+  using T = typename Container::value_type;
+  const int size = state.range(0);
+  for(auto _ : state)
+  {
+    Container arr;
+    arr.reserve(size);
+    assert_size_and_strict_capacity(arr, 0, size);
+    for(int i = 0; i < size; ++i)
+    {
+      arr.emplace_back(
+        get_value<T>(i));  // using emplace_back on container w/ reserved size
+    }
+    assert_size_and_strict_capacity(arr, size, size);
+    benchmark::DoNotOptimize(arr);
+  }
+}
+
+//-----------------------------------------------------------------------------
+// Use test fixtures for testing other operations
+//-----------------------------------------------------------------------------
+template <typename Container>
+struct ArrFixture
+{
+public:
+  using T = typename Container::value_type;
+
+  ArrFixture(int size)
+  {
+    data.resize(size);
+    for(int i = 0; i < size; ++i)
+    {
+      data.emplace_back(get_value<T>(i));
+    }
+  }
+
+  Container data;
+};
+
+template <typename Container>
+void iterate_range(benchmark::State& state)
+{
+  using T = typename Container::value_type;
+  const int size = state.range(0);
+  ArrFixture<Container> fixture(size);
+
+  const T element = get_value<T>(size / 2);
+  for(auto _ : state)
+  {
+    int count = 0;
+    for(const T& item : fixture.data)
+    {
+      if(item == element)
+      {
+        ++count;
+      }
+    }
+    assert(count == 1);
+    benchmark::DoNotOptimize(count);
+  }
+}
+
+template <typename Container>
+void iterate_direct(benchmark::State& state)
+{
+  using T = typename Container::value_type;
+  const int size = state.range(0);
+  ArrFixture<Container> fixture(size);
+
+  const T element = get_value<T>(size / 2);
+  for(auto _ : state)
+  {
+    int count = 0;
+    for(auto it = fixture.data.begin(); it != fixture.data.end(); ++it)
+    {
+      if(*it == element)
+      {
+        ++count;
+      }
+    }
+    assert(count == 1);
+    benchmark::DoNotOptimize(count);
+  }
+}
+
+//-----------------------------------------------------------------------------
+// Register all the tests
+//-----------------------------------------------------------------------------
+
+template <typename T>
+void RegisterBenchmark()
+{
+  auto tname = [](const std::string& n) {
+    return axom::fmt::format("{}<{}>", n, get_type_name<T>());
+  };
+
+  // clang-format off
+  if((args_benchmark_features & ArrayFeatureBenchmarks::Constructors) != ArrayFeatureBenchmarks::None)
+  {
+    benchmark::RegisterBenchmark(tname("Array::ctor"), &ctor<axom::Array<T>>)->Apply(CustomArgs);
```

Comment:
Taking the address of the functions surprised me here.

Reply:
I think it's because it's registering it at runtime, so needs a function pointer rather than using a macro (?)